src/mesh/domain.hpp

```diff
@@ -0,0 +1,194 @@
+    return dim <= interior_dims.size();
+  }
+
+  inline void MakeZeroDimensional_(int index) {
+    x_[index] = IndexRange(0, 0);
+    entire_ncells_[index] = 1;
+  }
+
+ public:
+  IndexShape() {}
+
+  IndexShape(const int &nx3, const int &nx2, const int &nx1, const int &ng)
+      : IndexShape(std::vector<int>{nx3, nx2, nx1}, ng) {}
+
+  IndexShape(const int &nx2, const int &nx1, const int &ng)
+      : IndexShape(std::vector<int>{nx2, nx1}, ng) {}
+
+  IndexShape(const int &nx1, const int &ng) : IndexShape(std::vector<int>{nx1}, ng) {}
+
+  //----------------------------------------------------------------------------------------
+  //! \fn IndexShape::IndexShape(const std::vector<int> & interior_dims, const int &ng)
+  //  \brief Builds IndexShape using interior_dims, note that the interior dims must
+  //  be specified in the order:
+  //
+  //  interior_dims.at(0) = nx3
+  //  interior_dims.at(1) = nx2
+  //  interior_dims.at(2) = nx1
+  //
+  IndexShape(std::vector<int> interior_dims, const int &ng) {
+    std::reverse(interior_dims.begin(), interior_dims.end());
+    assert(interior_dims.size() <= NDIM &&
+           "IndexShape cannot be initialized, the number of "
+           "dimensions exceeds the statically set dimensions, you will need to change "
+           "the NDIM "
+           "constant.");
+    for (int dim = 1, index = 0; dim <= NDIM; ++dim, ++index) {
+      if (DimensionProvided_(interior_dims, dim) == false) {
+        MakeZeroDimensional_(index);
+      } else {
+        assert(interior_dims.at(index) > -1 &&
+               "IndexShape cannot be initialized with a negative number of "
+               "interior cells for any dimension");
+        if (interior_dims.at(index) == 0) {
+          MakeZeroDimensional_(index);
+        } else {
+          x_[index] = IndexRange(ng, (ng + interior_dims.at(index) - 1));
+          entire_ncells_[index] = interior_dims.at(index) + 2 * ng;
+        }
+      }
+    }
+  }
+
+  std::array<IndexRange, NDIM> GetBounds(const IndexDomain &domain) const noexcept {
+    if (domain == IndexDomain::entire) {
+      std::array<IndexRange, NDIM> bounds;
+      for (int index = 0; index < NDIM; ++index) {
+        bounds[index].s = 0;
+        bounds[index].e = entire_ncells_[index];
+      }
+      return bounds;
+    }
+    return x_;
+  }
+
+  inline const IndexRange GetBoundsI(const IndexDomain &domain) const noexcept {
+    return (domain == IndexDomain::entire) ? IndexRange(0, entire_ncells_[0] - 1) : x_[0];
+  }
+
+  inline const IndexRange GetBoundsJ(const IndexDomain &domain) const noexcept {
+    return (domain == IndexDomain::entire) ? IndexRange(0, entire_ncells_[1] - 1) : x_[1];
+  }
+
+  inline const IndexRange GetBoundsK(const IndexDomain &domain) const noexcept {
+    return (domain == IndexDomain::entire) ? IndexRange(0, entire_ncells_[2] - 1) : x_[2];
+  }
+
+  inline int is(const IndexDomain &domain) const noexcept {
```

Comment:
These inline functions will need `KOKKOS_INLINE_FUNCTION` to be called on the GPU.